File: ml-parcial-IIND3118/ml-parcial-IIND3118/temas/A03_knn_y_kernel_regression/R/A03_knn_y_kernel_regression.R
# ############################################################################
# A3. KNN Y KERNEL
# Tema A03_knn_y_kernel_regression — Introducción al Machine Learning (IIND3118)
#
# CÓMO USARLO
#   1. Cambia el bloque marcado con ✏️ por tu dataset.
#   2. Corre:  Rscript A03_knn_y_kernel_regression.R     (o ábrelo en RStudio)
#   3. Las gráficas se guardan en la carpeta figuras/ de este tema.
#
# La explicación del método está en el README.md de esta misma carpeta.
# ############################################################################

for (p in c("MASS","cluster","dbscan","glmnet","leaps","caret","class","fpc",
            "clusterCrit","isotree","kernlab","Rtsne","psych","car","vegan")) {
  if (!requireNamespace(p, quietly=TRUE))
    try(install.packages(p, repos="https://cloud.r-project.org"), silent=TRUE)
}
suppressMessages({ library(MASS); library(cluster) })

set.seed(0)

# Las gráficas se guardan en figuras/ en vez de abrirse en pantalla
dir.create("../figuras", showWarnings=FALSE)
.n <- 0
guardar <- function() {
  .n <<- .n + 1
  dev.copy(png, sprintf("../figuras/%02d_figura.png", .n), width=900, height=650)
  invisible(dev.off())
}

# ╔═══════════════════════════════════════════════════════════════════════╗
# ║  ✏️  CAMBIA SOLO ESTE BLOQUE POR TU DATASET                           ║
# ╚═══════════════════════════════════════════════════════════════════════╝
DATOS  <- "../../../data/ejemplo_regresion.csv"
TARGET <- "target"        # nombre EXACTO de la columna respuesta. NULL si no hay.
# ╚═══════════════════════════════════════════════════════════════════════╝

df <- na.omit(read.csv(DATOS))

if (!is.null(TARGET)) {
  y_raw <- df[[TARGET]]
  X_df  <- df[, setdiff(names(df), TARGET), drop=FALSE]
} else { y_raw <- NULL; X_df <- df }

X_df <- as.data.frame(model.matrix(~ . - 1, data=X_df))
nombres_X <- names(X_df)
X  <- as.matrix(X_df)
Xs <- scale(X)

if (!is.null(y_raw)) {
  ES_CLASIFICACION <- is.factor(y_raw) || is.character(y_raw) || length(unique(y_raw)) <= 10
  y <- if (ES_CLASIFICACION) as.factor(y_raw) else as.numeric(y_raw)
  idx <- sample(seq_len(nrow(X)), floor(0.70*nrow(X)))
  Xtrain <- X[idx,]; Xtest <- X[-idx,]; ytrain <- y[idx]; ytest <- y[-idx]
  Xs_train <- Xs[idx,]; Xs_test <- Xs[-idx,]
} else { y <- NULL; ES_CLASIFICACION <- FALSE }

pca  <- prcomp(X, center=TRUE, scale.=TRUE)
Z2   <- pca$x[, 1:2]
col_pt <- if (!is.null(y)) as.integer(as.factor(y)) else "steelblue"

cat(strrep("=",74), "\n")
cat("A3. KNN Y KERNEL\n")
cat(strrep("=",74), "\n")
cat(sprintf("datos: %d obs. x %d predictores\n\n", nrow(X), ncol(X)))

# ############################################################################
#  CÓDIGO DEL MÉTODO
# ############################################################################

# ============================================================================
# A3. KNN Y KERNEL REGRESSION
# ----------------------------------------------------------------------------
# KNN, EN SIMPLE: para predecir un punto nuevo, busca los k puntos más
# parecidos y devuelve el PROMEDIO de sus y (regresión) o la clase más
# VOTADA (clasificación). No aprende ninguna fórmula: se guarda los datos.
#
# KERNEL REGRESSION: igual, pero en vez de "los k más cercanos" usa todos
# los que caigan dentro de una VENTANA de ancho h.
#   f(X*) = promedio de los Y_i cuyos X_i cumplen |X_i - X*| <= h
#
# ⚠️ HAY QUE ESTANDARIZAR (usa Xs): ambos usan distancias. Si una variable
#    está en millones y otra en decenas, la primera domina todo.
#
# PARÁMETROS
#   k    : el que se calibra. k PEQUEÑO = flexible. Prueba 1 a 40.
#   h    : ancho de ventana. h PEQUEÑO = flexible.
# ============================================================================

# --- KNN: calibrar k --------------------------------------------------------
errores <- numeric(40)
for (k in 1:40) {
  if (ES_CLASIFICACION) {
    pr <- knn(Xs_train, Xs_test, ytrain, k=k)      # class::knn
    errores[k] <- mean(pr != ytest)                # error de clasificación
  } else {
    pr <- predict(knnreg(Xs_train, ytrain, k=k), Xs_test)
    errores[k] <- mean((ytest - pr)^2)             # MSE
  }
}
k_opt <- which.min(errores)

plot(1:40, errores, type="b", pch=16, xlab="k (número de vecinos)",
     ylab=ifelse(ES_CLASIFICACION,"error de clasificación","MSE test"),
     main="Calibración de k")
abline(v=k_opt, col="red", lty=2)
cat(sprintf("k óptimo = %d\n", k_opt))

# --- Modelo final -----------------------------------------------------------
if (ES_CLASIFICACION) {
  pred_knn <- knn(Xs_train, Xs_test, ytrain, k=k_opt)
  cat(sprintf("Accuracy: %.4f\n", mean(pred_knn == ytest)))
} else {
  pred_knn <- predict(knnreg(Xs_train, ytrain, k=k_opt), Xs_test)
  mse_knn  <- mean((ytest - pred_knn)^2)
  cat(sprintf("MSE test: %.3f\n", mse_knn))
}

# --- Kernel regression: implementación de clase ----------------------------
kernelreg <- function(xstar, X_, Y_, h) {
  # h GRANDE = rígido (sesgo). h PEQUEÑO = flexible (varianza).
  sapply(xstar, function(xs) {
    dentro <- abs(xs - X_) <= h
    if (sum(dentro) == 0) 0 else mean(Y_[dentro])
  })
}

if (!ES_CLASIFICACION) {
  j     <- which.max(abs(cor(X, y)))     # variable más correlacionada con y
  x1    <- Xs[, j]
  malla <- seq(min(x1), max(x1), length.out=200)

  plot(x1, y, pch=16, cex=.5, col="gray",
       xlab=paste(nombres_X[j],"(estandarizada)"), ylab=TARGET,
       main="Kernel regression: efecto de h")
  for (hh in c(0.05, 0.3, 1.5)) {
    lines(malla, kernelreg(malla, x1, y, hh), lwd=2,
          col=c("green","blue","red")[match(hh, c(0.05,0.3,1.5))])
  }
  legend("topright", c("h=0.05 (varianza)","h=0.3","h=1.5 (sesgo)"),
         col=c("green","blue","red"), lwd=2)
}

cat("\nListo. Revisa la carpeta figuras/\n")

File: ml-parcial-IIND3118/ml-parcial-IIND3118/temas/A03_knn_y_kernel_regression/python/A03_knn_y_kernel_regression.py
"""
A3. KNN y kernel regression
Tema A03_knn_y_kernel_regression — Introducción al Machine Learning (IIND3118)

CÓMO USARLO
  1. Cambia el bloque marcado con ✏️ abajo por tu dataset.
  2. Corre el archivo:  python A03_knn_y_kernel_regression.py
  3. Las gráficas se guardan en la carpeta figuras/ de este tema.

La explicación del método está en el README.md de esta misma carpeta.
"""

import os, sys, warnings
warnings.filterwarnings("ignore")

import numpy as np
import pandas as pd
import matplotlib
matplotlib.use("Agg")          # no abre ventanas: funciona en cualquier entorno
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split, KFold, cross_val_score, GridSearchCV
from sklearn.preprocessing import scale
from sklearn.linear_model import (LinearRegression, Ridge, Lasso, ElasticNet,
                                  RidgeCV, LassoCV, ElasticNetCV, LogisticRegression)
from sklearn.neighbors import KNeighborsRegressor, KNeighborsClassifier
from sklearn.metrics import (mean_squared_error, mean_absolute_error, r2_score,
                             accuracy_score, confusion_matrix, classification_report,
                             roc_curve, roc_auc_score)
from sklearn.decomposition import PCA, KernelPCA
from sklearn.cluster import KMeans, DBSCAN
from sklearn.ensemble import IsolationForest
from sklearn.neighbors import LocalOutlierFactor
from sklearn import manifold

# ── Guardado automático de figuras ──────────────────────────────────────
AQUI = os.path.dirname(os.path.abspath(__file__))
FIGS = os.path.join(AQUI, "..", "figuras")
os.makedirs(FIGS, exist_ok=True)
_n = [0]
def _save():
    _n[0] += 1
    ruta = os.path.join(FIGS, f"{_n[0]:02d}_figura.png")
    plt.savefig(ruta, dpi=110, bbox_inches="tight"); plt.close()
    print(f"   [figura guardada: figuras/{os.path.basename(ruta)}]")
plt.show = _save                          # cada plt.show() guarda en vez de abrir

RNG = np.random.default_rng(0)

# ╔═════════════════════════════════════════════════════════════════════════╗
# ║  ✏️  CAMBIA SOLO ESTE BLOQUE POR TU DATASET                             ║
# ╚═════════════════════════════════════════════════════════════════════════╝
DATOS  = os.path.join(AQUI, "..", "..", "..", "data", "ejemplo_regresion.csv")
TARGET = "target"          # nombre EXACTO de la columna respuesta. None si no hay.
# ╚═════════════════════════════════════════════════════════════════════════╝

df = pd.read_csv(DATOS).dropna()

if TARGET is not None:
    y_serie = df[TARGET]; X_df = df.drop(columns=[TARGET])
else:
    y_serie = None; X_df = df.copy()

X_df = pd.get_dummies(X_df, drop_first=True).astype(float)
nombres_X = list(X_df.columns)
X  = X_df.values
Xs = scale(X)

if y_serie is not None:
    y = pd.factorize(y_serie)[0] if y_serie.dtype == object else y_serie.values
    ES_CLASIFICACION = (len(np.unique(y)) <= 10) and np.all(y == np.round(y))
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.30, random_state=0)
    Xs_train, Xs_test, _, _ = train_test_split(Xs, y, test_size=.30, random_state=0)
else:
    y = None; ES_CLASIFICACION = False

def ModeloLineal(**kw):
    return (LogisticRegression(max_iter=5000, **kw) if ES_CLASIFICACION
            else LinearRegression(**kw))

col = y if y is not None else "steelblue"
Z   = PCA().fit_transform(Xs)        # proyección PCA, se usa para graficar
Z2  = Z[:, :2]

print("="*74)
print("A3. KNN y kernel regression")
print("="*74)
print(f"datos: {X.shape[0]} obs. x {X.shape[1]} predictores")
if y is not None:
    print(f"y ('{TARGET}'): " + ("CLASIFICACIÓN" if ES_CLASIFICACION else "REGRESIÓN"))
print()

# ════════════════════════════════════════════════════════════════════════
#  CÓDIGO DEL MÉTODO
# ════════════════════════════════════════════════════════════════════════

# ═══ A3. KNN — calibrar k y evaluar ═══
# ⚠️ SIEMPRE con Xs (estandarizado), porque usa distancias.

Modelo = KNeighborsClassifier if ES_CLASIFICACION else KNeighborsRegressor

# --- 1. Calibrar k con la curva de test ---------------------------------
ks = range(1, 41)
errores = []
for k in ks:
    m = Modelo(n_neighbors=k).fit(Xs_train, y_train)
    p = m.predict(Xs_test)
    errores.append(1 - accuracy_score(y_test, p) if ES_CLASIFICACION
                   else mean_squared_error(y_test, p))

k_opt = list(ks)[int(np.argmin(errores))]

plt.plot(ks, errores, "o-", ms=3)
plt.axvline(k_opt, color="red", ls="--", label=f"k óptimo = {k_opt}")
plt.xlabel("k (número de vecinos)")
plt.ylabel("error de clasificación" if ES_CLASIFICACION else "MSE test")
plt.title("Calibración de k"); plt.legend(); plt.show()

# --- 2. Modelo final con el k calibrado ---------------------------------
knn = Modelo(
    n_neighbors=k_opt,      # ⬅ el k que calibraste
    weights="uniform",      # 'uniform' o 'distance'
).fit(Xs_train, y_train)

pred_knn = knn.predict(Xs_test)
mse_knn = mean_squared_error(y_test, pred_knn)
print(f"k óptimo = {k_opt}   |   MSE test = {mse_knn:.3f}")


# ═══ Kernel regression (ventana h) — implementación de clase ═══
def kernelreg(xstar, X_, Y_, h):
    """Promedio de los Y cuyos X caen dentro de [xstar-h, xstar+h].
       h GRANDE = modelo rígido (mucho sesgo). h PEQUEÑO = flexible."""
    pred = np.zeros(len(xstar))
    for i in range(len(xstar)):
        dentro = np.abs(xstar[i] - X_) <= h
        pred[i] = np.sum(np.where(dentro, Y_, 0)) / max(dentro.sum(), 1)
    return pred

# Demostración en 1 variable (la de mayor correlación con y)
j = int(np.argmax([abs(np.corrcoef(X[:, c], y)[0, 1]) for c in range(X.shape[1])]))
x1 = Xs[:, j]
orden = np.argsort(x1)
malla = np.linspace(x1.min(), x1.max(), 200)

plt.plot(x1, y, "o", ms=3, color="gray", alpha=.5, label="datos")
for h, col in [(0.05, "green"), (0.3, "blue"), (1.5, "red")]:
    plt.plot(malla, kernelreg(malla, x1, y, h), col, lw=2, label=f"h = {h}")
plt.xlabel(f"{nombres_X[j]} (estandarizada)"); plt.ylabel(TARGET)
plt.title("Kernel regression: efecto de h"); plt.legend(); plt.show()

print("h chico (verde) = persigue el ruido -> VARIANZA")
print("h grande (rojo) = casi una recta plana -> SESGO")

print("\nListo. Revisa la carpeta figuras/")
